neuraxle/checkpoints.py

```diff
@@ -19,185 +19,442 @@
+        Save the data container or fitted step checkpoint with the given data container, and context.
+        Returns the data container checkpoint, or latest data container.
 
-        :param expected_outputs: initial expected outputs of pipeline to load checkpoint from
-        :param data_inputs: data inputs to save
-        :return: self
+        :param data_container: data container to save data container or fitted steps
+        :param context: context to save data container or fitted steps
+        :return: saved data container
+        :rtype: DataContainer
         """
-        return self
+        raise NotImplementedError()
 
-    def transform(self, data_inputs):
+    @abstractmethod
+    def read_checkpoint(self, data_container: DataContainer, context: ExecutionContext) -> DataContainer:
         """
-        Save checkpoint for data inputs and expected outputs so that it can
-        be loaded by the checkpoint pipeline runner on the next pipeline run
-
-        :param data_inputs: data inputs to save
-        :return: data_inputs
+        Read the data container checkpoint with self.data_checkpointer.
+        Returns a new data container loaded with all the data inputs,
+        and expected outputs for each current id in the given data container.
+
+        :param data_container: data container containing the current_ids to read checkpoint for
+        :param context: context to read checkpoint for
+        :return: the data container checkpoint
+        :rtype: DataContainer
         """
-        return data_inputs
+        raise NotImplementedError()
+
+class StepCheckpointer(BaseCheckpointer):
+    """
+    StepCheckpointer is used by the Checkpoint step to save the fitted steps contained in the context of type ExecutionContext.
+
+    By default, StepCheckpointer saves the fitted steps when the execution mode is either FIT, or FIT_TRANSFORM :
+    ```
+    StepCheckpointer(ExecutionMode.FIT_OR_FIT_TRANSFORM)
+
+    # is equivalent to :
+
+    StepCheckpointer()
+    ```
+    """
+    def __init__(
+        self,
+        execution_mode: ExecutionMode = ExecutionMode.FIT_OR_FIT_TRANSFORM,
+    ):
+        BaseCheckpointer.__init__(self, execution_mode=execution_mode)
+
+    def save_checkpoint(
+            self,
+            data_container: DataContainer,
+            context: ExecutionContext
+    ) -> DataContainer:
+        # TODO: save the context by execution mode AND data container ids / summary
+        context.save_all_unsaved()
+        return data_container
+
+    def should_resume(self, data_container: DataContainer, context: ExecutionContext) -> bool:
+        # TODO: change this when we support multiple execution modes and data container ids / summary
+        return True
+
+
+class BaseDataContainerSaver(ABC):
 
     @abstractmethod
-    def set_checkpoint_path(self, path):
+    def get_checkpoint_file_path(self, checkpoint_path: str, current_id) -> str:
```

Comment:
Remember that some checkpoints might be in the cloud and not with a strictly-speaking "path". I think something is wrong here.

Reply:
what do you want to be passed as argument ? do you only want a rename ? removed file from the name of the method.

Reply:
@guillaume-chevalier ?